### DISCUSSION TITLE
Can't make action using docker and calling python script work

### DISCUSSION THREAD

**Discussion by HorstBaerbel** [community/community#26515] [Actions]
<p>Hi All.</p>
<p>I’m trying to make an action to run clang-format on my repository. I have used all of the components successfully before in a workflow, except the docker part. Now I wanted to <a href="https://github.com/HorstBaerbel/action-clang-format" rel="nofollow noopener">create an action</a> I can reuse, but when using the action in the workflow of another repository the “clang-format” run tells me “/usr/bin/env: ‘python’: No such file or directory” (run results [here](http:///usr/bin/env:%20’python’: No such file or directory)) or the docker COPY command fails and it fails to find the run-clang-format.py file. Copying entrypoint.sh seems to work though, as it can be run…</p>
<p>What I really wanted to achieve is:</p>
<ul>
<li>
<p>Either a “reusable” workflow yaml file I can send extra parameters to (using: workflow/bla@master, with: …).</p>
</li>
<li>
<p>Or an action that does NOT need all the npm or docker cruft and simply runs a shell script (using: “my custom shell command”) on my repo, just like workflow files can do already.</p>
</li>
</ul>
<p>Any ideas on how to fix the docker action bug and / or alternatively achieve what I want?</p>

**Comment by airtower-luna** [comment]
<p>The “No such file or directory” error sounds like a combination of two problems:</p>
<ol>
<li>The “python” binary on Debian and related distros (including Ubuntu) is Python 2, which has been deprecated for a long time and is out of support since start of this year. Use python3 instead.</li>
<li>Python (any version) likely isn’t installed in the base ubuntu:latest container (I know it isn’t in the debian containers). Try adding “python3” to the apt install step in your Dockerfile after fixing the point above.</li>
</ol>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/horstbaerbel">@horstbaerbel</a> , </p>
<p>I totally agree with airtower-luna. You could install python3 in your dorkerfile.</p>
<pre><code>RUN apt-get update \
  &amp;&amp; apt-get install -y python3-pip python3-dev \
  &amp;&amp; cd /usr/local/bin \
  &amp;&amp; ln -s /usr/bin/python3 python \
  &amp;&amp; pip3 install --upgrade pip
</code></pre>
<p>Please see my changed dockerfile: </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/13a80b9d3aefe89bbca54b1cff942865ce9ef75e.png" data-download-href="/uploads/short-url/2NT6F9cvFLoptU9BMhPjFyj5EC2.png?dl=1" title="dockerfile example.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181098118-15c96321-cc80-4400-8fe2-31d66abbc132.png" alt="dockerfile example.png" data-base62-sha1="2NT6F9cvFLoptU9BMhPjFyj5EC2" width="621" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/13a80b9d3aefe89bbca54b1cff942865ce9ef75e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">dockerfile example.png</span><span class="informations">691×556 13.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>  </p>

**Comment by HorstBaerbel** [comment]
<blockquote>
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> wrote:</p>
</blockquote>
<blockquote>
<p>The “No such file or directory” error sounds like a combination of two problems:</p>
</blockquote>
<blockquote>
<ol>
<li>The “python” binary on Debian and related distros (including Ubuntu) is Python 2, which has been deprecated for a long time and is out of support since start of this year. Use python3 instead.</li>
<li>Python (any version) likely isn’t installed in the base ubuntu:latest container (I know it isn’t in the debian containers). Try adding “python3” to the apt install step in your Dockerfile after fixing the point above.</li>
</ol>
</blockquote>
<p>Reading the output it seems like python was already being installed, but there was maybe no link python3 -&gt; python:</p>
<pre><code>The following additional packages will be installed:
  file libclang-cpp9 libedit2 libexpat1 libllvm9 libmagic-mgc libmagic1
  libmpdec2 libpython3-stdlib libpython3.6-minimal libpython3.6-stdlib
  libreadline7 libsqlite3-0 libssl1.1 mime-support python3 python3-minimal
  python3.6 python3.6-minimal readline-common xz-utils
</code></pre>
<p>I use:</p>
<pre><code>RUN apt update
RUN apt -y install clang-format-9 python3
RUN update-alternatives --install /usr/bin/clang-format clang-format /usr/bin/clang-format-9 100
RUN update-alternatives --install /usr/bin/python python /usr/bin/python3 100
</code></pre>
<p>now. And it works! Thanks people! That was rather simple…</p>
<p>A couple of more question:</p>
<ul>
<li>
<p>It takes time to build the docker image. Is there a way to cache and re-use it for other actions / workflows?</p>
</li>
<li>
<p>Is there an option / docker FROM command to use the GitHub linux runner docker image? This has most of what I need (except clang-tools) already installed. Or do you know of a docker image with clang-tools installed? I found <a href="https://github.com/conan-io/conan-docker-tools" rel="nofollow noopener">the conan docker tools</a>, but am not sure how to use the images.</p>
</li>
<li>
<p>Is there an option to NOT run a heavy js/npm or docker command in the action, but just  a simple shell script like in the workflow file which can just run a python script inside the “run:” command? Obviously then there’s no way to copy files to the workspace, but I don’t understand why workflow files support this, but actions don’t…</p>
</li>
</ul>
<p>A example use case is that I would like to make a “build” action (see workflow file <a href="https://github.com/HorstBaerbel/res2h/blob/master/.github/workflows/build.yml" rel="nofollow noopener">here</a>) that runs cmake + make that I can reuse in other repos. Or an action that runs CTest - all simple shell commands. If I use an action I have to use a docker file and have to build everything from scratch again, which is really slow…</p>
<p>And thanks again for the help!</p>

**Comment by HorstBaerbel** [comment]
<p>To answer part of my own question:</p>
<blockquote>
<ul>
<li>It takes time to build the docker image. Is there a way to cache and re-use it for other actions / workflows?</li>
</ul>
</blockquote>
<p>I have built a docker image containing all necessary programs <a href="https://hub.docker.com/repository/docker/horstbaerbel/ccpp-cmake-build-and-test" rel="nofollow noopener">here</a>. This should be more efficient than installing everything each run…</p>
